tests/topotests/bgp_vpnv4_allocnh/test_bgp_vpnv4_allocnh.py

```diff
@@ -0,0 +1,292 @@
+            assert nh["type"] == "BGP", "{}, show mpls table, nexthop is not installed".format(router.name)
+            if "nexthop" in nh.keys():
+                assert nh["nexthop"] not in nexthopList, "{}, show mpls table, duplicate nexthop address".format(router.name)
+                nexthopList.append(nh["nexthop"])
+            elif "interface" in nh.keys():
+                assert nh["interface"] not in nexthopList, "{}, show mpls table, duplicate nexthop interface".format(router.name)
+                nexthopList.append(nh["interface"])
+            else:
+                assert 0, "{}, show mpls table, entry with neither nexthop nor interface detected".format(router.name)
+
+def test_protocols_convergence():
+    """
+    Assert that all protocols have converged
+    statuses as they depend on it.
+    """
+    tgen = get_topogen()
+    if tgen.routers_have_failure():
+        pytest.skip(tgen.errors)
+
+    # Check BGP IPv4 routing tables on VRF1 of r1
+    logger.info("Checking BGP IPv4 routes for convergence on r1 VRF1")
+    router = tgen.gears['r1']
+    json_file = "{}/{}/bgp_ipv4_routes_vrf1.json".format(CWD, router.name)
+    if not os.path.isfile(json_file):
+        assert 0, 'bgp_ipv4_routes_vrf1.json file not found'
+
+    expected = json.loads(open(json_file).read())
+    test_func = partial(
+        topotest.router_json_cmp,
+        router,
+        "show bgp vrf vrf1 ipv4 json",
+        expected,
+    )
+    _, result = topotest.run_and_expect(test_func, None, count=20, wait=0.5)
+    assertmsg = '"{}" JSON output mismatches'.format(router.name)
+    assert result is None, assertmsg
+
+    logger.info("Checking BGP VPNv4 routes for convergence on r2")
+    router = tgen.gears['r2']
+    json_file = "{}/{}/bgp_vpnv4_routes.json".format(CWD, router.name)
+    if not os.path.isfile(json_file):
+        assert 0, 'bgp_vpnv4_routes.json file not found'
+
+    expected = json.loads(open(json_file).read())
+    test_func = partial(
+        topotest.router_json_cmp,
+        router,
+        "show bgp ipv4 vpn json",
+        expected,
+    )
+    _, result = topotest.run_and_expect(test_func, None, count=10, wait=0.5)
+    assertmsg = '"{}" JSON output mismatches'.format(router.name)
+    assert result is None, assertmsg
+
+    # Check BGP labels received on r2
+    logger.info("Checking BGP VPNv4 labels on r2")
+    labelList = set()
+    bgp_vpnv4_table_check(tgen.gears["r2"], labelList)
+
+    # Check MPLS labels received on r1
+    mpls_table_check(tgen.gears['r1'], labelList)
+
+def test_flapping_bgp_vrf():
+    """
+    Turn down a remote BGP session
+    """
+    tgen = get_topogen()
+    if tgen.routers_have_failure():
+        pytest.skip(tgen.errors)
+    logger.info("Turning off r11-eth0 on r11")
+    tgen.gears['r11'].link_enable('r11-eth0', enabled=False)
+    topotest.sleep(5, "waiting that BFD event propagates")
```

Comment:
No sleep everywhere, please.

Reply:
Even more, bfd is not relevant here at all?

Reply:
BFD helps me to converging BGP to down faster.
I will check if sleep can be removed.
I don't see the point in removing BFD. is it forbidden to use BFD in tests?

Reply:
ok. removed